src/client/Microsoft.Identity.Client/Utils/OptionalSemaphoreSlim.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Utils
+{
+    /// <summary>
+    /// An object that either wraps a SemaphoreSlim for syncronization or ignores syncronization completely and just keeps track of Wait / Release operations.
+    /// </summary>
+    internal class OptionalSemaphoreSlim
```

Comment:
This "adapter" object maintains the interface of SemaphoreSlim, while allowing for 

CC @trwalke for a example of a pattern, roughly resembling adapter?